File: 4.OOPinJava/4.4.Singleton/Logger.java
// Task 4:
// Create a Singleton class `Logger` ensuring only one instance exists.
// Provide a `log(String message)` method that prefixes messages with a timestamp.

import java.time.LocalDateTime;

class Logger {
    private static Logger instance;
    private Logger() { }

    // TODO: implement getInstance()

    public void log(String message) {
        System.out.println("[" + LocalDateTime.now() + "] " + message);
    }
}

public class Main {
    public static void main(String[] args) {
        Logger log1 = Logger.getInstance();
        Logger log2 = Logger.getInstance();

        log1.log("First message");
        log2.log("Second message");
        System.out.println(log1 == log2);  // Expected: true
    }
}

File: 3.OOPinPython/3.4.Singleton/logger.py
"""
Task 4:
Create a Singleton class `Logger` ensuring only one instance exists.
Provide a `log(message)` method that prefixes messages with a timestamp.
"""

from datetime import datetime

class Logger:
    # TODO: add class variable for instance

    # implement constructor

    def log(self, message: str) -> None:
        # TODO: implement log method
        pass

def main():
    log1 = Logger()
    log2 = Logger()

    log1.log("First message")
    log2.log("Second message")
    print(log1 is log2)  # Expected: True

if __name__ == "__main__":
    main() 